app_memory: Cleanup of the interface.

AdithyaBaglody opened the pull request:
This patch makes it easy for the users to use this feature.
Also the userspace test case is modified to use these new
interfaces.

Apart from the macros that were added for easy usage of this feature. A few bug were found and were fixed in these patches.

    The app shared memory feature was built to read linker variables
    and then update the size of various regions at runtime.
    Due to the optimizations of the C language and the linker these
    variables were not reflecting the correct values. Therefore the
    size of the bss and the data regions if not correctly configured
    would cause a kernel oops.

    This python script will create sections and linker variables to place the
    application shared memory partitions.
    These are later read by the macros defined in app_memdomain.h for
    initialization purpose when APP_SHARED_MEM is enabled.
    This new feature will reduce the compiled code and will decrease
    the boot time.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
app_memory: Cleanup of the interface.

This patch makes it easy for the users to use this feature.
Also the userspace test case is modified to use these new
interfaces.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented on subsys/app_memory/app_memdomain.c:
```diff
@@ -92,6 +92,19 @@ void app_calc_size(void)
+	 * For CONFIG_ARC_MPU_VER == 3, these partitions are not added
+	 * into MPU now, so need to disable mpu first to do app_bss_zero()
+	 */
+	arc_core_mpu_disable();
```
@ruuddw @vonhust  I am confused here. This code is supposed to run in supervisor mode. But still I see that mpu needs to be disabled to access these memory regions. I don't understand why this needs to be here.
Can you please help me in understanding this.

andrewboie replied:
might need to change the perms on the memory regions so that supervisor mode has full access, what's the ARC memory domain map flags look like?

ruuddw replied:
Good suggestion to move the 'disable' code from the test application to the generic subsys code. Indeed ARC initial memory map prohibits access to the appmem addresses, it's only enabled on switching to usermode, I believe. So for this bss initialization in between initial MPU setup and switching to userspace, MPU protection is disabled.

ioannisg replied:
>Good suggestion to move the 'disable' code from the test application to the generic subsys code. Indeed ARC initial memory map prohibits access to the appmem addresses, it's only enabled on switching to usermode, I believe. So for this bss initialization in between initial MPU setup and switching to userspace, MPU protection is disabled.

Could we add some of this text in the comment, to make it more clear why we need to disable and enable?

AdithyaBaglody replied:
>Good suggestion to move the 'disable' code from the test application to the generic subsys code. Indeed ARC initial memory map prohibits access to the appmem addresses, it's only enabled on switching to usermode, I believe. So for this bss initialization in between initial MPU setup and switching to userspace, MPU protection is disabled

oh okay. I will add this in the code as comments. Will help later on.

AdithyaBaglody committed:
app_memory: Added new macros for ease of use.

This patch ensures smaller, easier macros for the user.

The current approach expects the user to have the all the
operations to be carried out in a single file. Which doesn't
usually happen. This was seriously restricting the usage of
this feature. Now these functions can be called by any
function/file which has the correct permissions.

The following are the new macros:
1. APPMEM_PARTITION_OBJECT_DEFINE to create objects in a .c files.
2. APPMEM_PARTITION_HEADER_DEFINE will create static inline
   functions and extern definitions.
3. APPMEM_DOMAIN_OBJECT_DEFINE for creating objects
4. APPMEM_DOMAIN_HEADER_DEFINE for creating static inline functions
   and extern definitions.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented:
I think I need to see these new APIs in action to properly review, might be good to include the test case changes in the PR.

ioannisg commented on include/app_memory/app_memdomain.h:
```diff
@@ -82,32 +82,52 @@ struct app_region {
+	FOR_EACH(appmem_partition_header, __VA_ARGS__)
+
+/* Returns the mem_partition struct */
+#define APPMEM_PARTITION_GET(name) (mem_partition_##name)
```
Should this return the pointer?

ioannisg replied:
I could not find where this is used, actually

AdithyaBaglody replied:
> Should this return the pointer?

I guess you are right we should return a pointer.

>I could not find where this is used, actually

i will be pushing a new PR soon with a few test cases that have been ported to use these APIs.

andrewboie replied:
were you going to make this change?

ioannisg commented on include/app_memory/app_memdomain.h:
```diff
@@ -82,32 +82,52 @@ struct app_region {
-		mem_domain_##name.start = (u32_t) name.dmem_start; \
-		mem_domain_##name.attr = K_MEM_PARTITION_P_RW_U_RW; \
-		name.partition = &mem_domain_##name; \
+		mem_partition_##name.start = (u32_t) name.dmem_start; \
```
That's nice

ioannisg commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -33,14 +33,23 @@ K_SEM_DEFINE(expect_fault_sem, 0, 1);
-FOR_EACH(appmem_partition, part0, part1, part2);
+APPMEM_PARTITION_OBJECT_DEFINE(part0, part1, part2);
 
+/* this defines all the functions that are needed to configure
```
This

ioannisg commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -33,14 +33,23 @@ K_SEM_DEFINE(expect_fault_sem, 0, 1);
-FOR_EACH(appmem_domain, dom0, dom1);
+APPMEM_DOMAIN_OBJECT_DEFINE(dom0, dom1);
+
+/* this defines all the functions that are needed to configure
```
This

ioannisg reviewed (commented):
In general, i liked this cleanup.
Left some comments.
@AdithyaBaglody  please, fix CI :)

AdithyaBaglody commented:
>I think I need to see these new APIs in action to properly review, might be good to include the test case changes in the PR.

These are getting used in test cases like mem_protect/mem_protect. I want to have a clear segregation between these 2 changes. I will push a new PR to depict the usecases for these new APIs soon.

AdithyaBaglody committed:
kernel: init: Use linker defined variable for size calculation

Use of linker variable _app_smem_size will optimize the code.
This will also eliminate any optimization related problems.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
scripts: gen_app_partitions: Update to support all architectures.

The app shared memory feature was built to read linker variables
and then update the size of various regions at runtime.
Due to the optimizations of the C language and the linker these
variables were not reflecting the correct values. Therefore the
size of the bss and the data regions if not correctly configured
would cause a kernel oops.

So this patch will make this script generic enough to be able to
run on all architectures that support app shared memory.

This script will create sections and linker variables to place the
application shared memory partitions.
These are later read by the macros defined in app_memdomain.h for
initialization purpose when APP_SHARED_MEM is enabled.
This new feature will reduce the compiled code and will decrease
the boot time.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
arch: x86: linker.ld: app_shared_mem will now use generated linker file.

The app shared memory feature was built to read linker variables
and then update the size of various regions at runtime.
Due to the optimizations of the C language and the linker these
variables were not reflecting the correct values. Therefore the
size of the bss and the data regions if not correctly configured
would cause a kernel oops.

This patch will make x86 always take up the generated linker file.
Which fixes a lot of issues with the alignment between partitions
etc.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
arch: arm: linker: app_shared_mem will now use generated linker file.

The app shared memory feature was built to read linker variables
and then update the size of various regions at runtime.
Due to the optimizations of the C language and the linker these
variables were not reflecting the correct values. Therefore the
size of the bss and the data regions if not correctly configured
would cause a kernel oops.

This patch will make ARM devices always take up the generated
linker file. Previously this was only enabled for power of 2 MPUs.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
arch: arc: linker: app_shared_mem will now use generated linker file.

The app shared memory feature was built to read linker variables
and then update the size of various regions at runtime.
Due to the optimizations of the C language and the linker these
variables were not reflecting the correct values. Therefore the
size of the bss and the data regions if not correctly configured
would cause a kernel oops.

This patch will make ARC devices always take up the generated
linker file. Previously this was only enabled for power of 2 MPUs.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
x86: crt0.S: Data copy is needed for app_shared_mem.

This is needed becuase we have initialized data which needs to
goto the RAM.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
app_memdomain: Optimize the app_shared_mem feature.

The app shared memory feature was built to read linker variables
and then update the size of various regions at runtime.
Due to the optimizations of the C language and the linker these
variables were not reflecting the correct values. Therefore the
size of the bss and the data regions if not correctly configured
would cause a kernel oops.

The python script gen_app_partitions.py will create sections and
linker variables to place the application shared memory partitions.
These are later read by the macros defined in app_memdomain.h for
initialization purpose when APP_SHARED_MEM is enabled.

This patch will read these linker defined variables and configure
the required structures. This enabled in the removal of the API
app_calc_size() from app_memdomain.c
This new feature will reduce the compiled code and will decrease
the boot time.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
CMakeLists: Allow APP_SHARED_MEM to run for all supported architectures

This patch will run the python scripts for all architectures
to determine the partitions available. This is needed for correct
operation of the app_shared_mem feature.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
CMakeLists: APP_SHARED_MEM runs after all the libs are built.

This particular rule of cmake was causing problems when run in
multi-threaded environment. This change in the dependency will
resolve all such issues and ensures that this script is executed
correctly.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
app_shared_mem: Support libc in applications.

This consists of 2 changes.
1. The sections are now created with _data and _bss respectively.
   This will make the linker script a lot cleaner and easy to
   understand.
2. The script is updated to reflect point 1 and also additional
   code has been added to support libc.

Now data and bss of libc comes in as a partition, which can be
configured like any other partitions.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
ztest: Update the ztest framework for app_shared_mem.

This patch enables the usage of app_shared_mem for all the tests.
The patch introduces a new Kconfig for ztest called
CONFIG_ZTEST_INIT_APP_SHARED_MEM. When disabled the application
will take the onus of initializing all the partitions and domains.

This patch makes it easy for any test case to be ported to shared
mem. The application will have to only mark their data and bss
regions by macro APP_DATA_MEM / APP_BSS_MEM. This will push the
required variables into a specific memory from where they will be
configured for k_mem_partition.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
samples: userspace: shared_mem: Updated the sample.

Updated the sample to reflect the new interface of app shared
memory.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: Updated few testcases for APP_SHARED_MEM.

This patch adds required macros to place data/bss memory in
correct memory locations.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: timing_info: Updated to use APP_SHARED_MEM.

This benchmark doesn't use ztest framework therefore we have
to create new partitions. Into which we had to add the required
variables. This is mainly needed for userspace benchmarks.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: kernel: mem_pool: sys_mem_pool: Updated for APP_SHARED_MEM.

The mem pool should be placed in a region accessible by the test
cases. In this case the mem pool is made to reside in the data region
of ztest partition.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: kernel: mem_protect: Updated to work with APP_SHARED_MEM.

Updated the userspace test cases to work with shared mem.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: kernel: thread_apis: Updated to work with APP_SHARED_MEM.

Updated the thread_apis test cases to work with shared mem.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: kernel: thread_init: Updated for APP_SHARED_MEM.

The test thread_init creates 2 threads statically. In such a
scenario we need to assign the domains to these threads as soon as
the test_main starts.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: kernel: Updated for APP_SHARED_MEM.

The following tests have been updated for APP_SHARED_MEM.
1. msgq/msgq_api
2. pipe/pipe
3. pipe/pipe_api
4. workq/work_queue
5. queue

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: kernel: stack: Updated to work with APP_SHARED_MEM.

Updated the stack/stack_api and stack/stack_usage test cases
to work with shared mem.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: cmsis_rtos_v1: Updated for APP_SHARED_MEM.

There was an issue with the reuse of the thread id. This patch
now ensures that the problematic thread id is killed properly.
Also for userspace this testcase was generating far too many
kernel objects.Hence bumped up the thread bytes from 2 to 3.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: kernel: schedule_api: Updated to work with APP_SHARED_MEM.

Updated the test cases to work with shared mem.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: posix: Increase the max thread bytes.

The number of kobjects that were getting created exceeded the normal
2 byte limit. This patch now increase this from 2 byte to 4.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: net: socket: getaddrinfo: Updated for APP_SHARED_MEM.

This test case uses libc from user threads. In order to support
libc, a partition was created and initialized. This is also
supported by the gen_app_partitions.py script. The script will
push all the required data/bss to the correct location inside the
partition. Thus supporting libc from user threads.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: Kconfig: Move from APPLICATION_MEMORY to APP_SHARED_MEM.

The APPLICATION_MEMORY is now superseded by APP_SHARED_MEM.
All the userspace applications will now use APP_SHARED_MEM.
Therefore we will use the same in the tests.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: cmsis_rtos_v2: Updated for APP_SHARED_MEM.

For userspace this testcase was generating far too many
kernel objects.Hence bumped up the thread bytes from 2 to 3.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: mutex: Increased delay.

Random crashes on the CI. Hence increased the delay because app
shared mem might contribute extra time for boot up.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: schedule_api: Increased the slice time for calculation.

When the app shared memory was turned on, the time slice calculations
were incorrect. This patch increased the duration from 200ms to 300ms.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented:
@andrewboie please review.

AdithyaBaglody committed:
tests: kernel: common: Increased the timer duration.

The k_timer was not allowing enough time for the test case
timeout_order to complete correctly.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

pfalcon commented:
Was it really required to patch many seemingly unrelated tests in this PR? For example, I apparently was added as a reviewer because the patch includes tests/net/socket/getaddrinfo/src/main.c. But that test is supposed to test getaddrinfo() call and nothing else.

pfalcon reviewed (commented):
Was it really required to patch many seemingly unrelated tests in this PR? For example, I apparently was added as a reviewer because the patch includes tests/net/socket/getaddrinfo/src/main.c. But that test is supposed to test getaddrinfo() call and nothing else.

AdithyaBaglody commented:
@pfalcon 
>Was it really required to patch many seemingly unrelated tests in this PR

This feature is getting enabled for all usersapce supporting platforms. The test case you are referring to was causing some problem with application shared memory. Not just this but a lot of the test cases were touched.

andrewboie commented on subsys/app_memory/app_memdomain.c:
```diff
@@ -92,6 +92,21 @@ void app_calc_size(void)
+	 * For CONFIG_ARC_MPU_VER == 3, these partitions are not added
+	 * into MPU now, so need to disable mpu first to do app_bss_zero()
+	 */
+	extern void arc_core_mpu_disable(void);
```
Ugh.
Is there any way we can do this without hacking in ARC-specific calls into what is supposed to be platform-agnostic subsystem code?

andrewboie commented on include/app_memory/app_memdomain.h:
```diff
@@ -82,36 +82,52 @@ struct app_region {
  */
 #define appmem_domain(name) \
-	__kernel struct k_mem_domain domain_##name; \
+	__kernel struct k_mem_domain domain_##name;
```
why does this have a trailing semicolon?

andrewboie commented on include/app_memory/app_memdomain.h:
```diff
@@ -82,36 +82,52 @@ struct app_region {
 
 /* A helper function that should be called to define multiple partitions*/
-#define APPMEM_PARTITION_DEFINE(...)\
+#define APPMEM_PARTITION_OBJECT_DEFINE(...)\
```
Please fix your git history.
You defined this to one thing in the previous patch, now you are changing it

andrewboie commented on kernel/init.c:
```diff
@@ -188,7 +188,7 @@ void _data_copy(void)
 #endif	/* CONFIG_CODE_DATA_RELOCATION */
 #ifdef CONFIG_APP_SHARED_MEM
 	(void)memcpy(&_app_smem_start, &_app_smem_rom_start,
-		 ((u32_t) &_app_smem_end - (u32_t) &_app_smem_start));
```
It's not clear to me that this optimizes anything, wouldn't the compiler do this arithmetic at build time anyway?

AdithyaBaglody replied:
GCC was doing something very strange, the smem_end and smem_start value changes with new objects getting added. But it was not getting reflected when the code was run. This was a strange situation. hence i had to use a  "_size"  liker variable to get this done.

andrewboie commented on include/arch/x86/linker.ld:
```diff
@@ -202,25 +202,27 @@ SECTIONS
+#define SMEM_PARTITION_ALIGN(size) MMU_PAGE_ALIGN
+#define APP_SHARED_ALIGN  MMU_PAGE_ALIGN
+
+#include<app_smem.ld>
```
need a space before <

andrewboie commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -411,22 +411,9 @@ SECTIONS
-		 */
-		APP_SHARED_ALIGN;
-		_app_smem_end = .;
+#define SMEM_PARTITION_ALIGN MPU_ALIGN
```
Unless I'm missing something you are now enforcing power of two alignment on all ARM targets

AdithyaBaglody replied:
No not really . The alignement is done by a macro out side app_smem.ld. If we have a power of 2 MPU the alignment computed is different compared to a NXP MPU.

nashif reviewed (changes requested):
Your commit message needs more details and all those macros being added to the headers need to end up in our API documentation somehow. Where is this feature being documented and why I do not see any change to any documentation?

> This patch makes it easy for the users to use this feature

Please explain how is this making it easy

andrewboie commented on CMakeLists.txt:
```diff
@@ -1163,30 +1163,29 @@ configure_file(
+    -o ${APP_SMEM_LD}
+    $<$<BOOL:${CMAKE_VERBOSE_MAKEFILE}>:--verbose>
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}/
+    COMMENT "Generating power of 2 aligned app_smem linker section"
```
Is this still true? We are no longer behind if(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)

AdithyaBaglody replied:
Nope this is no longer valid. comment is wrong. will fix this.

andrewboie commented on tests/ztest/Kconfig:
```diff
@@ -47,3 +47,11 @@ config ZTEST_PARAMETER_COUNT
 	help
 	  Maximum amount of concurrent return values / expected parameters.
+
+config ZTEST_INIT_APP_SHARED_MEM
```
Ugh, I don't like that this config needs to exist.
We really need initialization of domains/partitions to be automatic at boot, like we do with statically defined kernel objects, and the only manual calls that need to be made are to assign threads to them.
I guess this can be addressed later.

AdithyaBaglody replied:
By deafult this is set to true. for instance if someone wants to have a test case in which they would like to manually do smem init, they can set this to No. This is just a extension people can use later.

andrewboie commented on tests/kernel/mem_pool/sys_mem_pool/src/main.c:
```diff
@@ -21,9 +21,13 @@
 K_MUTEX_DEFINE(pool_mutex);
+#ifdef CONFIG_APP_SHARED_MEM
+SYS_MEM_POOL_DEFINE(pool, &pool_mutex, BLK_SIZE_MIN, BLK_SIZE_MAX,
+		    BLK_NUM_MAX, BLK_ALIGN, data_smem_ztest_part0_data);
```
There's got to be a better way to do this, maybe a helper macro for the section name?

AdithyaBaglody replied:
Nope not possible. the macro SYS_MEM_POOL_DEFINE will take the last arguement as a sting. 
if we add a macro gcc will not preprocess and add it.

andrewboie commented on tests/kernel/mem_protect/mem_protect/src/mem_domain.c:
```diff
@@ -305,8 +317,8 @@ K_MEM_PARTITION_DEFINE(mem_domain_tc3_part8_struct,
 	&mem_domain_tc3_part1_struct,
-	&mem_domain_tc3_part2_struct,
 	&mem_domain_tc3_part3_struct,
+	&mem_domain_tc3_part2_struct,
```
Why was the ordering changed here?

andrewboie commented on tests/kernel/threads/thread_init/src/main.c:
```diff
@@ -205,6 +205,12 @@ void test_kinit_coop_thread(void)
 {
+#ifdef CONFIG_APP_SHARED_MEM
+	/* app shared memory needs to be assigned to the following 2 threads */
+	appmem_add_thread_ztest_dom0(T_KDEFINE_COOP_THREAD);
```
It would be nice if we could direct this at toplevel, like we already do with K_THREAD_ACCESS_GRANT()

andrewboie commented on tests/kernel/queue/src/test_queue_contexts.c:
```diff
@@ -14,20 +14,20 @@ K_QUEUE_DEFINE(kqueue);
-static qdata_t data_p[LIST_LEN];
-static qdata_t data_l[LIST_LEN];
-static qdata_t data_sl[LIST_LEN];
+APP_BSS_MEM struct k_queue queue;
```
Why is a kernel object going into app memory?

andrewboie replied:
Actually, why was this file changed at all?
All the tests in here run in supervisor mode, it's only test_queue_user.c that has user mode tests

andrewboie commented on tests/kernel/queue/src/test_queue_contexts.c:
```diff
@@ -14,20 +14,20 @@ K_QUEUE_DEFINE(kqueue);
 static K_THREAD_STACK_DEFINE(tstack1, STACK_SIZE);
-static struct k_thread tdata1;
-static struct k_sem end_sema;
+APP_BSS_MEM static struct k_thread tdata1;
```
Why is a kernel object going into app memory?

andrewboie commented on tests/kernel/queue/src/test_queue_loop.c:
```diff
@@ -10,13 +10,13 @@
-static qdata_t data_p[LIST_LEN];
-static qdata_t data_r[LIST_LEN];
-static struct k_queue queue;
+APP_BSS_MEM static qdata_t data[LIST_LEN];
```
None of these changes should be necessary, and we should not be putting kernel objects in application memory anyway

andrewboie commented on tests/kernel/sched/schedule_api/src/test_priority_scheduling.c:
```diff
@@ -27,8 +27,8 @@ static K_SEM_DEFINE(sema2, 0, NUM_THREAD);
-static int thread_idx;
-static struct k_thread t[NUM_THREAD];
+static APP_BSS_MEM int thread_idx;
+static APP_BSS_MEM struct k_thread t[NUM_THREAD];
```
Do not put kernel objects in app memory

andrewboie replied:
Why did you even change the schedule_api test? It does not run in user mode...

andrewboie commented on tests/kernel/sched/schedule_api/src/test_sched_is_preempt_thread.c:
```diff
@@ -12,8 +12,8 @@
-static struct k_thread tdata;
-static struct k_sem end_sema;
+static APP_BSS_MEM struct k_thread tdata;
+static APP_BSS_MEM struct k_sem end_sema;
```
Do not put kernel objects in app memory

andrewboie commented on tests/kernel/sched/schedule_api/src/test_sched_priority.c:
```diff
@@ -7,8 +7,8 @@
 static K_THREAD_STACK_DEFINE(tstack, STACK_SIZE);
-static struct k_thread tdata;
-static int last_prio;
+static APP_BSS_MEM struct k_thread tdata;
```
Do not put kernel objects in app memory

andrewboie commented on tests/kernel/sched/schedule_api/src/test_sched_timeslice_and_lock.c:
```diff
@@ -9,16 +9,16 @@
-static struct k_thread tthread[THREADS_NUM];
-static int old_prio, init_prio;
+static APP_BSS_MEM struct thread_data tdata[THREADS_NUM];
+static APP_BSS_MEM struct k_thread tthread[THREADS_NUM];
```
Do not put kernel objects in app memory

tyrantlizard77 reviewed (changes requested):
You seem to have changed a bunch of tests which don't even run in user mode. And you are doing things like placing kernel objects in app memory which is definitely incorrect.
Please go through and check that your changes are actually necessary

andrewboie commented:
Also where are the changes to doc/ since you modified the API?

AdithyaBaglody commented:
@andrewboie all the tests that were modified were failing when APP_SMEM was made default with userspace. Therefore i fixed them to get them working with APP_SMEM.
